### DISCUSSION TITLE
Code editing improvement ideas welcome!

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#261] [General]
I'm finishing up the code-editor improvements this coming week, as we need to now focus on the next stage.

If you can think of any little things, no matter how small, that would help with your code editing experience with tB, do please open an issue feature request for consideration.

The new 'paste as comment' option is a perfect example of something easy to implement, yet really helps provide for a better coding experience.

Have a nice weekend folks!

**Comment by cristianbuse** [comment]
1. It would be nice to be able to generate getters/setters for a class member variable.

2. I sometime am lazy and type in VBA something like:  
```VBA
function SomeFunc as string
```
which gets prettified as:  
```VBA
Function SomeFunc() As String

End Function
```
meanwhile in TB is gets changed to:  
```VBA
Function SomeFuncasstring()
    
End Function
```
would be nice if TB would recognize the missing parenthesis.

3. Would be nice if I could automatically generate all the implemented member functions for an implemented interface. Something similar with how VB* does when we choose an entry in the top-right section of a code window.

**Reply by A9G-Data-Droid** [reply]
For 1. I would prefer auto-properties that support .NET style flags like `ReadOnly` to show a property that only has public getter. No private backing variable required. It would have a private setter and public getter by default. If you added a getter line then you could choose to return something else, instead of the implicit private backer.

**Comment by A9G-Data-Droid** [comment]
I would like some sort of intellicode suggestions that help us discover new twinBasic language features. To avoid slowing the live intellisense maybe it could be a set of code Analyzers like Roslyn. You could import a VBA project, run code inspections, and twinBasic could show you all of the new things it can do to simplify your code with the new language features. This would be a great starting point for anyone wanting to begin porting their old code to learn what's new. It should accept plugin sets so you could add community developed code inspections, like RubberDuck.

**Reply by WaynePhillipsEA** [reply]
Yes, this is pretty much what #51 will be for 👍

**Comment by A9G-Data-Droid** [comment]
`Go to Definition` for `GoTo` statements. When I right click the name of a `GoTo` label I get a tooltip that says "No definition found for `labelName`"

I expect this to take me to the target of the `GoTo`, the line label. This would allow you to navigate spaghetti faster.

**Reply by WaynePhillipsEA** [reply]
Yes, good catch, thanks

**Comment by A9G-Data-Droid** [comment]
**Auto-Indent Module:**  There should be a way to fix indenting on an imported module that was ugly when imported. I would expect to find it by right clicking in the code window.

**Reply by WaynePhillipsEA** [reply]
Yes this is planned.  It will work similarly to Ruberduck/Smart Indenter via a 'format section' / 'format document' menu option

**Comment by DavidHollman** [comment]
I always found the need for line continuation characters _ to be quite tedious. Making them optional (but I guess still accepted for backwards compatibility) would be an improvement IMO.

**Reply by WaynePhillipsEA** [reply]
Good idea @DaveInCaz.  I know most most usage of line continuation characters are optional in VB.NET, so it is definitely worth consideration.  Could you open a separate issue to discuss?

**Reply by DavidHollman** [reply]
Done! https://github.com/WaynePhillipsEA/twinbasic/issues/282

**Comment by DavidHollman** [comment]
Building up a string from a long series of concatenations could be more efficient (and less error prone) if there were some type of TB-native string substitution syntax.

C# uses an "interpolation" syntax like `$"text {variable} more text"` which I find works very well. Curiously despite the use of `$` it doesn't seem to be particularly "BASIC"-ish syntax which should necessarily just be copied. I'm not sure if VB.Net or some other Basic variant has something similar that would be a good fit for the language?

**Reply by XusinboyBekchanov** [reply]
https://docs.microsoft.com/en-us/dotnet/visual-basic/programming-guide/language-features/strings/interpolated-strings

**Reply by WaynePhillipsEA** [reply]
Yes, string interpolation is planned.  Discussed briefly in #132

**Reply by WaynePhillipsEA** [reply]
I personally quite like the back-tick character for string interpolation, as seen in JavaScript (and probably other languages).  But that's up for discussion.

**Reply by mansellan** [reply]
Personally, I prefer `$` - it's more canonical to BASIC:
- In VBx, it's used as a string hint, 
- In VB.Net (and C#), it's used as an interpolation directive.
-  In other BASICs it's the only identifier of a string.
 
It has history.

Backticks have never signified anything special in any BASIC dialog I can think of.

**Reply by bclothier** [reply]
I agree with mansellan - `$"..."` and `@"..."` and of course, the combined `$@"..."` seems to be the best syntax. I know BASIC isn't supposed to use symbols but this seems to me the least worst choice compared to alternatives. One more consideration -- BASIC uses `'` to mean a comment or a string. Allowing backticks to have meaning would be quite taxing on eyes.... was that a quote or a backtick? Gotta squint at it...

**Reply by WaynePhillipsEA** [reply]
Well, with semantic highlighting we can give different types of strings different colours, so there would be no squinting 😜

Though all other points duly noted.

**Comment by pvb-mike** [comment]
I could not find a way to customize the colors used for syntax highlighting. This is something I would really like.

I struggle to look at my source code with the default color selection. All those different colors now make it appear so distracting that I just cannot use the current twinBASIC IDE.

I think many other people with attention deficit problems as well as people with color blindness will find the ability to customize the colors a necessity.

Perhaps a quick solution could be a single option to toggle on/off all syntax highlighting?

**Comment by FullValueRider** [comment]
In the twinbasic folder, in the IDE subfolder you will find
• styles.css
• styles-twinbasic-dark.css
• styles-twinbasic-light.css

If you edit these files you can change colours in the ide.  save copies outside of the twinbasic installation folder as these files get overwritten when you upgrade to the latest beta version.

**Reply by WaynePhillipsEA** [reply]
On a related note, when updating to a new BETA, *do not* overwrite the new ZIP contents over an existing IDE folder.  Always extract to a fresh clean folder as mixing files with an old version is known to have caused weird issues in the past.

**Comment by rexxitall** [comment]
Seems someones is looking for work :) Ok 
Most used tools by me in the last years is the code indenter also for the whole project
The indenter shall stop when he find a error but shall on whish show his current result (makes it easy to find the mistake)
a cross reference lister
a tool to automatically create class setter and getter,
all was MZ tools has to offer should be build in,
input und print with color in the output console,
build in terminal for the command line
build in KI tool for chatgpt and google AI

something weired for mixed programming would also be nice.
a easy way to call any kind of dll functions ( managed, unmanaged, c, fortran styled, .net ) without the hussle to convert data types.

easy interprocess communication with PIPES, tcp, messages, websockets ...

just as suggestions 
it is what i currently use while programming

I hope something is in the box for you and your team :)

Have fun
Thomas
